File: problem2.java
// Time Complexity : O(n * amount)
// Space Complexity : O(n * amount) we use 2-D DP array of the size
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// For each coin, the algorithm fills up the dp matrix iteratively, considering
// the combinations needed to reach each amount using the current coin and the previously calculated combinations.
class problem2 {
    public int change(int amount, int[] coins) {
        int n = coins.length;
        int[][] dp = new int[n + 1][amount + 1];
        for (int i = 0; i < n; i++) {
            dp[i][0] = 1;
        }

        for (int i = 1; i <= amount; i++) {
            dp[0][i] = 0;
        }

        for (int i = n - 1; i >= 0; i--) {
            for (int j = 1; j <= amount; j++) {
                if (coins[i] > j) {
                    dp[i][j] = dp[i + 1][j];
                } else {
                    dp[i][j] = dp[i + 1][j] + dp[i][j - coins[i]];
                }
            }
        }

        return dp[0][amount];
    }
}

File: problem2.go

// Time Complexity : o(m*n)---> O(N)
// Space Complexity : O(m*n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

func change(amount int, coins []int) int {
    if len(coins) == 0{
        return 0
    }
    
    
    m := len(coins) //rows
    n := amount //columns
 
    
    dp := make([][]int,m+1)//rows
    for i,_ := range(dp){
        dp[i] = make([]int,n+1)//columns
    }
    
    for i:=0;i<m+1;i++{
        dp[i][0] = 1
    }
    
    for i:=1;i<m+1;i++{
        for j:=1;j<n+1;j++{
            if j < coins[i-1]{
                dp[i][j] = dp[i-1][j]
            }else{
                dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]]
            }
            
        }
    }
    
    return dp[m][n]
}
